Blocks: Try new context API to simplify display of block controls

gziolo opened the pull request:
## Description
This PR updated React version to version 16.3.x (once it is stable we can update https://github.com/WordPress/gutenberg/commit/15cd537054fd227c0281341d77a9fe52a8ebb2ed and move to its own PR). One of the new features that 16.3 is going to introduce is [new context API](https://medium.com/@baphemot/whats-new-in-react-16-3-d2c9b7b6193b). I gave it a spin to simplify the way we handle `<InspectorControls />` and `<BlockContols />` inside `edit` component included in every block.

### Before

```
function edit( { isSelected } ) {
    return [
        isSelected && (
            <BlockControls key="controls">
                ...
            </BlockControls>
        ),
        isSelected && (
           <InspectorControls key="inspector">
                ...
            </InspectorControls>
        ),
        <RichText
                tagName="p"
                isSelected={ isSelected }
             ...
         />,
    ];
}
```

### After

```
function edit( { isSelected } ) {
    return [
        <BlockControls key="controls">
            ...
        </BlockControls>,
        <InspectorControls key="inspector">
            ...
        </InspectorControls>,
        <RichText
                tagName="p"
             ...
         />,
    ];
}
```

### Reasoning

I never liked that we need to prepend those controls with the check for every block. It is very easy to forget about it and it leads to unexpected behaviors. It turns out that this is a quite common issue that people have when learning how to create a new block. This is what I understood when talking to @zgordon. He shared also an example issue where the current logic caused troubles: [Duplicate Inspector Controls](https://javascriptforwp.com/forums/topic/duplicate-inspector-controls/).

### Implementation

This PR wraps `<EditBlock />` component, which renders `edit` component for each block, with its own context which stores `isSelected` value inside a consumer for later use. Both controls (`<InspectorControls />` and `<BlockContols />`) get wrapped with a context consumer which renders them only when `isSelected` is truthy. As simple as that.
The same technique was used for `<RichText />` component and its `< BlockFormatControls />` after suggestion from @youknowriad .

## How Has This Been Tested?

Manually tested if there were no regression. The only block that was upgraded so far is `<Paragraph />` to use this new context. If we agree that this is the change we want, I'm happy to update all existing blocks (as a follow-up) ~~and the documentation~~ (docs updated).

The best way to test is to add a few paragraphs and make sure the display only their own controls.

I also made sure all unit tests pass (`npm run test`) and e2e tests pass on Travis.

## TODO

* [ ] Add deprecation notice to inform that `isSelected` is no longer mandatory.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

gziolo committed:
Blocks: Refactor block controls to use block edit context

youknowriad commented:
I really like the idea, this change in addition to the removal of the focus management will make block author's life easier.

I'm always suspicious when it comes to context usage in React in general but I think this is a good usage for it. `BlockControls` and `InspectorControls` slots are already magic, so adding a context value to simplify their usage makes sense to me especially since this check is consistent across all blocks.

gziolo committed:
Docs: Update docs to reflect the change in how block controls are handled

gziolo commented:
I figured out that build is failing because Enzyme doesn't support new tags created by `createContext`. See related issue: https://github.com/airbnb/enzyme/issues/1509.

gziolo commented:
Updated with the recent changes in `master` and updated to React 16.3.0-rc.0.

Should we also do a similar trick for `<RichText />`? In that case, it might require a more sophisticated logic because in some places we pass some overrides based on other conditions.

gziolo commented on lib/client-assets.php:
```diff
@@ -379,16 +379,16 @@ function gutenberg_register_vendor_scripts() {
 	gutenberg_register_vendor_script(
 		'react',
-		'https://unpkg.com/react@16.2.0/umd/react' . $react_suffix . '.js'
+		'https://unpkg.com/react@16.3.0-alpha.1/umd/react' . $react_suffix . '.js'
```
I missed that, it needs to be updated...

youknowriad commented on blocks/block-controls/index.js:
```diff
@@ -3,11 +3,18 @@
+ */
+import { ifEditBlockSelected } from '../block-edit/context';
+
+export function BlockControls( { controls, children } ) {
```
How can we inform the user about this backward compatibity change. I think it's fine to ship it because users are probably already testing `isSelected` on their side but how can we inform them that it's not useful anymore?

gziolo replied:
It isn't a breaking change, so it's more about the convenience of developers and assuring that they don't introduce subtle bug as noted in [here](https://javascriptforwp.com/forums/topic/duplicate-inspector-controls/). I don't think we can add a deprecation message in this case as it's passed down in case of `BlockControls` and `InspectorControls`.

youknowriad replied:
I was thinking maybe not a warning but we add a log message like this when registering custom blocks:

```
In this new Gutenberg Release, The `BlockControls`,  `InspectorControls` and `RichText` automatically hides when the block is not selected, you don't need to check or pass the `isSelected` prop anymore.
```

And just leave this message for one release at least? What do you think?

gziolo replied:
Yes, I like this idea. I didn't think of it this way :)

youknowriad reviewed (commented):
I believe we should push this further and update the default value of the `isSelected` prop of the `RichText` component to be `true` now and wrap the Formatting Toolbar of the RichText component with this new HoC as well.

mcsf commented:
I like where this is going. It feels like a good balance of _magic_, and the reason why it doesn't go overboard is that the abstraction itself makes sense: being in a _context_ of block editing is something that makes sense for the block developer and for the user, and the idea that `BlockControls` _et al._ should be dependent on this context makes total sense.

> I believe we should push this further and update the default value of the isSelected prop of the RichText component to be true now and wrap the Formatting Toolbar of the RichText component with this new HoC as well.

👍

zgordon commented:
I like this continued direction towards automagical and the notice in the console is something most would pick up. Nice to not be breaking change for those who started using isSelected 👍

gziolo committed:
Blocks: Add isSelected handling to InspectorAdvancedControls

gziolo committed:
Blocks: Replace getWrapperDisplayName with its new alternative

gziolo committed:
Blocks: Use createSlotFill for BlockControls

gziolo committed:
Blocks: Introduce BlockFormatControls component

gziolo committed:
Blocks: Make sure that RichText controls are only displayed when block is selected
This makes it backward compatible with the existing core blocks

gziolo commented on blocks/library/paragraph/test/__snapshots__/index.js.snap:
```diff
@@ -3,31 +3,25 @@
+            aria-multiline="false"
+            class="wp-block-paragraph blocks-rich-text__tinymce"
+            contenteditable="true"
+            data-is-placeholder-visible="false"
```
This is very unexpected. It looks like context provider doesn't work properly with Enzyme, but I can't observe the same in the browser. Very annoying thing :(

gziolo commented:
@youknowriad I addressed your comment related to `RichText`. It should work as expected without `isSelected` prop passed. I still need to add deprecation message to inform plugin developers about changes introduced.

Enzyme update is necessary to make tests work properly. I had to skip some tests and one of the tests behaves strangely (the one for Paragraph).

gziolo commented on blocks/block-controls/index.js:
```diff
@@ -1,13 +1,25 @@
+ */
+import { ifBlockEditSelected } from '../block-edit/context';
+
+const Fill = createSlotFill( 'BlockControls' );
```
I have an improvement for this one in another PR:
https://github.com/WordPress/gutenberg/pull/6086/files#diff-f55821c6d9b2038d9e462f87c1c9e0feR6

I will extract and update this PR if it is going to be ready to merge sooner.

youknowriad commented on blocks/block-edit/context.js:
```diff
@@ -0,0 +1,63 @@
+ */
+import { createContext, createHigherOrderComponent } from '@wordpress/element';
+
+const { Consumer, Provider } = createContext( {
```
I feel this context could also include what's current in the `BlockEdit` component's context right ? BlockList and canUserUseUnfilteredHTML

gziolo replied:
Yes, we can consolidate it in this PR using new Context API unless it breaks more tests ...

gziolo replied:
`BlocksList` looks complex because it depends on `createInnerBlockList` injected from `editor` ...
Another reason to merge those files to `editor` module :)

gziolo committed:
Blocks: Update test helper to work with block edit context

gziolo committed:
Blocks: Make sure that EditContextProvider is only rerendered when isSelected changes

gziolo commented on blocks/block-edit/context.js:
```diff
@@ -16,11 +16,36 @@ const { Consumer, Provider } = createContext( {
-			<OriginalComponent { ...props } />
-		</Provider>
-	);
+	return class ComponentWithProvider extends Component {
```
See https://reactjs.org/docs/context.html#caveats for more details.

> Because context uses reference identity to determine when to re-render, there are some gotchas that could trigger unintentional renders in consumers when a provider’s parent re-renders.

gziolo commented:
> I feel this context could also include what's current in the BlockEdit component's context right ? BlockList and canUserUseUnfilteredHTML

@youknowriad - I believe this should be done independently. It's quite complex logic and there are some implications involved. I think we might want to move some components to `editor` module first to simplify some code.

youknowriad commented on blocks/block-edit/context.js:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @return {Component} Component with a BlockEdit context set.
+ */
+export const withBlockEditContextProvider = createHigherOrderComponent( ( OriginalComponent ) => {
```
I don't like the fact that Provider is included using a HoC instead of a regular Component using children. Because it gives the impression that we could use this component in several places where in reality the providers are used only once.

Thoughts?

gziolo replied:
Good point made :)
I will move it to `BlockEdit`.

> Because it gives the impression that we could use this component in several places where in reality the providers are used only once.

Well, not in every case, but in general I agree 👍

youknowriad reviewed (approved):
LGTM 👍 I believe a second opinion would be good here

gziolo committed:
Blocks: Merge Provider with BlockEdit

gziolo committed:
Edit Post: Add information for developers about `isSelected` changes introduced

gziolo commented:
Added message on the JS console to let developer know about changes introduced:

![screen shot 2018-04-13 at 15 30 00](https://user-images.githubusercontent.com/699132/38737489-aeda6ba8-3f2f-11e8-87d6-3884fb89e491.png)

gziolo commented:
> LGTM 👍 I believe a second opinion would be good here

@youknowriad thanks for the review. @mcsf or @aduth would you mind confirming it is good to go?

aduth commented on blocks/block-edit/index.js:
```diff
@@ -19,8 +19,18 @@ import {
 export class BlockEdit extends Component {
+	constructor( props ) {
+		super( props );
+		this.state = {
```
I don't think a component needs _both_ the initial setting of state _and_ `static getDerivedStateFromProps`. The latter supersedes the need for the former. The constructor could be dropped altogether.

https://codepen.io/aduth/pen/QmRZLO

**Edit:** That said, it appears to be needed if you plan to `return null;` from `static getDerivedStateFromProps`, since otherwise the initial value of `this.state` is `undefined`.

https://codepen.io/aduth/pen/NYVOPa

aduth replied:
Since `getDerivedStateFromProps` is _always_ called on the initial mount, do we really need to set the values of state in the constructor? Or can we just set it to an empty object?

gziolo replied:
Yes, let's do that. We will probably have to use `get` to check state, but in overall it's still better.

gziolo replied:
Much better: https://github.com/WordPress/gutenberg/pull/5029/commits/fbb66c8052514e13b077ccaa51f09541d934e8f9

aduth commented on blocks/block-edit/context.js:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @return {Component} Component which renders only when the BlockEdit is selected.
+ */
+export const ifBlockEditSelected = createHigherOrderComponent( ( OriginalComponent ) => {
```
I like this 👍

aduth commented on edit-post/index.js:
```diff
@@ -57,6 +57,15 @@ export function initializeEditor( id, post, settings ) {
 	const reboot = reinitializeEditor.bind( null, target, settings );
 
+	if ( 'production' !== process.env.NODE_ENV ) {
+		// Remove with 3.0 release.
```
I wonder if we could implement this logging with an object property getter on a block's `this.props.isSelected`, so we're not feeling compelled to remove it soon, and to not log when the message is not applicable (could be confusing, since a developer might assume there's something actively wrong with a block, even if there's not).

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get

gziolo replied:
My main concern is that often this is used as follows:
```js
isSelected && (	
	<InspectorControls key="inspector">
		...
	</InspectorControls>     
)
```

but also appears in a different context:
```js
{ images.map( ( img, index ) => (
	<li className="blocks-gallery-item" key={ img.id || img.url }>
		<GalleryImage
			url={ img.url }
			alt={ img.alt }
			id={ img.id }
			isSelected={ isSelected && this.state.selectedImage === index }
			onRemove={ this.onRemoveImage( index ) }
			onSelect={ this.onSelectImage( index ) }
			setAttributes={ ( attrs ) => this.setImageAttributes( index, attrs ) }
			caption={ img.caption }
		/>
	</li>
) ) }
```

It is very hard to handle it on a case by case basis.

aduth commented on editor/components/block-toolbar/index.js:
```diff
@@ -18,8 +18,8 @@ function BlockToolbar( { block, mode } ) {
 			<BlockSwitcher uids={ [ block.uid ] } />
-			<Slot name="Block.Toolbar" />
-			<Slot name="Formatting.Toolbar" />
+			<BlockControls.Slot />
```
Glad we're losing the magic strings 👍

aduth commented on docs/deprecated.md:
```diff
@@ -9,6 +9,7 @@ Gutenberg's deprecation policy is intended to support backwards-compatibility fo
 ## 2.7.0
 
 - `wp.element.getWrapperDisplayName` function removed. Please use `wp.element.createHigherOrderComponent` instead.
+- `isSelected` usage is no longer mandatory with `BlockControls`, `InspectorControls` and `RichText`. It is now handled by the editor internally to ensure that controls are visible only when block is selected. See updated docs: https://github.com/WordPress/gutenberg/blob/master/blocks/README.md#components.
```
This is a bit out of place; the document is grouped by versions where specific features are removed, as a hint to developers to ensure updating in time. In this case, there's nothing backwards-incompatible.

gziolo replied:
Well, yes. It's tricky. I will remove and ask @mtias to include in the release post.

aduth commented on blocks/rich-text/index.js:
```diff
@@ -819,16 +821,16 @@ export class RichText extends Component {
-					</Fill>
-				}
-				{ isSelected && inlineToolbar &&
+				{ isSelected && ! inlineToolbar && (
```
Is `isSelected` here redundant? Since `BlockFormatControls` is already wrapped with `ifBlockEditSelected`.

__Edit:__ I see we have specific handling of `isSelected` in this component. What's the use case for that?

youknowriad replied:
The usecase is for multiple RichText blocks (quote for example)

aduth replied:
Confusing (two tiers of `isSelected`), but okay 😄

gziolo replied:
Yes, multiple RichText is quite difficult to handle and maintain. I will look into it next week. Maybe it wouldn't be that hard to automatically handle active state in a way where only one RichText can be active per block at a time.

gziolo replied:
If we would have `isBlockSelected` and `isRichTextSelected` it would be easier to read.

jorgefilipecosta replied:
> Maybe it wouldn't be that hard to automatically handle active state in a way where only one RichText can be active per block at a time.

I think we can have only one active RichText for the whole post. Maybe we can just use a component id and store id of RichText that is active, it would simplify the logic of some blocks.

jayshenk replied:
I like the idea of storing a single active RichText in the editor store, if this is what you mean. It would be useful for https://github.com/WordPress/gutenberg/pull/5794, so that inline images could be inserted anywhere RichText is used (not just in the Paragraph block), and so that the global inserter could select the state of the active RichText in the editor store when determining whether to render the Inline Blocks menu. CC @nb

gziolo replied:
> I think we can have only one active RichText for the whole post. Maybe we can just use a component id and store id of RichText that is active, it would simplify the logic of some blocks.

Yes, that's true. In addition, there might be only one active RichText per selected block. So I suppose we can reuse Context API for the same purpose and store the unique id of the actually active RichText component to make sure we don't display controls for the remaining components.

aduth commented on blocks/library/paragraph/index.js:
```diff
@@ -224,8 +226,6 @@ class ParagraphBlock extends Component {
 					</PanelBody>
 				</InspectorControls>
-			),
-			<div key="editable">
```
This appears useless.. but are we sure it was?

gziolo replied:
I can revert, but I didn't discover any issues so far in my testing. There are 5ish more div wrappers there 🤣

gziolo replied:
I will revert this particular change to be on the safe side, but I will try to remove it anyway in a separate PR when tackling this nested `div`s hell :)

gziolo committed:
Blocks: Improve state handling for BlockEdit context

gziolo committed:
Blocks: Addressed feedback related to BlockEditContext changes

gziolo commented:
Moving forward with this one. There is no easy way to provide a message to the developers that `isSelected` in some cases is optional that's why I proceeded with the `console.info` approach. I will share on Twitter and #core-editor channel about this change. That's the only thing we can do.
